Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,509 @@
+        var isClosed: Atomic<Bool>
+        /// The curently opened, unleased connections in this connection provider
+        var availableConnections: CircularBuffer<Connection> = .init(initialCapacity: 8) {
+            didSet {
+                self.closeIfNeeded()
+            }
+        }
+
+        /// Consumers that weren't able to get a new connection without exceeding
+        /// `maximumConcurrentConnections` get a `Future<Connection>`
+        /// whose associated promise is stored in `Waiter`. The promise is completed
+        /// as soon as possible by the provider, in FIFO order.
+        var waiters: CircularBuffer<Waiter> = .init()
+
+        var leased: Int = 0 {
+            didSet {
+                self.closeIfNeeded()
+            }
+        }
+
+        let maximumConcurrentConnections: Int = 8
+
+        let lock = Lock()
+
+        unowned let parentPool: ConnectionPool
+
+        init(group: EventLoopGroup, key: ConnectionPool.Key, configuration: HTTPClient.Configuration, initialConnection: Connection, parentPool: ConnectionPool) {
+            self.loopGroup = group
+            self.configuration = configuration
+            self.key = key
+            self.availableConnections.append(initialConnection)
+            self.parentPool = parentPool
+            self.isClosed = Atomic(value: false)
+        }
+
+        deinit {
+            assert(availableConnections.isEmpty, "Available connections should be empty before deinit")
+            assert(leased == 0, "All leased connections should have been returned before deinit")
+        }
+
+        /// Get a new connection from this provider
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            return self.lock.withLock {
+                if leased < maximumConcurrentConnections {
+                    leased += 1
+                    if let connection = availableConnections.popFirst(), connection.channel.isActive {
+                        connection.isLeased = true
+                        return self.loopGroup.next().makeSucceededFuture(connection)
+                    } else {
+                        return self.makeConnection(preference: preference)
+                    }
+                } else {
+                    let promise = self.loopGroup.next().makePromise(of: Connection.self)
+                    self.waiters.append(Waiter(promise: promise, preference: preference))
+                    return promise.futureResult
+                }
+            }
+        }
+
+        private func makeConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            let bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: self.loopGroup, host: self.key.host, port: self.key.port, configuration: self.configuration) { channel in
+                channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+            }
```

Comment:
There's definitely some repeated code between here and the initial connection setup: it'd be really good to factor those in a way that shows their common threads.

Reply:
This is still true, there's a lot of similar code between these two places.

Reply:
No more duplication thanks to https://github.com/swift-server/async-http-client/pull/105/commits/5d3b0cc34d31e02dad6b9b2c7a9bc7de48b2abbf as @artemredkin suggested 🙂